formula/src/main/java/com/instacart/formula/internal/SynchronizedEventQueue.kt

```diff
@@ -0,0 +1,74 @@
+package com.instacart.formula.internal
+
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * A non-blocking event queue that processes formula updates.
+ */
+class SynchronizedEventQueue {
+    private val threadRunning = AtomicReference<Thread>()
+    private val concurrentLinkedQueue = ConcurrentLinkedQueue<() -> Unit>()
+
+    /**
+     * All top-level formula interactions that trigger formula side-effects are posted here
+     * to make sure that they are executed one at a time. If there is a thread currently running
+     * formula, we hand the update to that thread for processing. The following
+     * root formula events are propagated via this queue:
+     * - Input change
+     * - Individual formula transitions
+     * - Termination
+     *
+     * Implementation works by having a concurrent queue and checking:
+     *  - If queue is idle, execute current update and try to process other queue entries
+     *  - If queue is running by the same thread, we execute current update and let other
+     *  updates be handled by existing processing loop.
+     *  - If queue is running by a different thread, add to the queue and see if we need to
+     *  take over the processing.
```

Comment:
something to consider is the cost of context switching if we rapidly switch for small evaluation tasks. It might make sense to batch things if this can be the case

Reply:
Could you explain more about what you mean by context switching?

Reply:
Thread context switching is the process of saving the state of a currently running thread and then restoring the state of another thread, allowing it to resume execution. This essentially allows multiple threads to share a single CPU, giving the illusion that they are running concurrently. The cost is fairly low for IO threads but high for computation threads

Reply:
We are trying to avoid thread-switching and enable running the formula on the thread that the event is produced on. Currently, all network calls switch to the main thread and then process the event on that thread. In the future, we will not switch to the main thread, but process the event first and only switch to the main thread when updating the UI.

Reply:
Also, we are using a non-blocking (atomic) approach to take over processing so the penalty should be relatively small.

Reply:
ok that makes sense